node/src/components/linear_chain/pending_signatures.rs

```diff
@@ -77,7 +109,33 @@ impl PendingSignatures {
         let validator_sigs = self.pending_finality_signatures.get_mut(public_key)?;
         let sig = validator_sigs.remove(&block_hash);
         self.remove_empty_entries();
-        sig
+        sig.map(|status| status.into_inner())
+    }
+
+    /// Marks pending finality signature as created by a bonded validator.
+    /// Returns whether there was a signature from the `public_key` for `block_hash` in the
+    /// collection.
+    pub(super) fn mark_bonded(&mut self, public_key: PublicKey, block_hash: BlockHash) -> bool {
+        self.mark_bonded_helper(public_key, block_hash).is_some()
+    }
+
+    fn mark_bonded_helper(&mut self, public_key: PublicKey, block_hash: BlockHash) -> Option<()> {
+        match self.pending_finality_signatures.entry(public_key.clone()) {
+            Entry::Occupied(mut validator_sigs) => {
+                let sig = validator_sigs.get_mut().remove(&block_hash)?;
+                if sig.is_bonded() {
+                    return Some(());
```

Comment:
Doesn't this effectively remove signatures that were already marked as bonded? Is that what we want? If yes, it would be good to document this behavior.

Reply:
It's not :grimacing: Good catch!

Reply:
Fixed in [`92fc19d` (#1291)](https://github.com/CasperLabs/casper-node/pull/1291/commits/92fc19df709560bd676ca39e826997f3c11a10ae). Thanks for pointing it out!